engine/apps/mobile_app/tasks.py

```diff
@@ -153,78 +184,189 @@ def _get_fcm_message(alert_group, user, registration_id, critical):
             },
         ),
     )
+
+    return _construct_fcm_message(device_to_notify, thread_id, fcm_message_data, apns_payload, critical)
+
+
+def _get_youre_going_oncall_fcm_message(
+    user: User, schedule: OnCallSchedule, device_to_notify: FCMDevice, seconds_until_going_oncall: int
+) -> Message:
+    thread_id = f"{schedule.public_primary_key}:{user.public_primary_key}:going-oncall"
+    data: FCMMessageData = {
+        "title": f"You are going on call in {humanize.naturaldelta(seconds_until_going_oncall)} for schedule {schedule.name}",
+    }
+
+    return _construct_fcm_message(device_to_notify, thread_id, data)
+
+
+@shared_dedicated_queue_retry_task(autoretry_for=(Exception,), retry_backoff=True, max_retries=MAX_RETRIES)
+def notify_user_async(user_pk, alert_group_pk, notification_policy_pk, critical):
+    # avoid circular import
+    from apps.base.models import UserNotificationPolicy, UserNotificationPolicyLogRecord
+
+    try:
+        user = User.objects.get(pk=user_pk)
+    except User.DoesNotExist:
+        logger.warning(f"User {user_pk} does not exist")
+        return
+
+    try:
+        alert_group = AlertGroup.all_objects.get(pk=alert_group_pk)
+    except AlertGroup.DoesNotExist:
+        logger.warning(f"Alert group {alert_group_pk} does not exist")
+        return
+
+    try:
+        notification_policy = UserNotificationPolicy.objects.get(pk=notification_policy_pk)
+    except UserNotificationPolicy.DoesNotExist:
+        logger.warning(f"User notification policy {notification_policy_pk} does not exist")
+        return
+
+    def _create_error_log_record():
+        """
+        Utility method to create a UserNotificationPolicyLogRecord with error
+        """
+        UserNotificationPolicyLogRecord.objects.create(
+            author=user,
+            type=UserNotificationPolicyLogRecord.TYPE_PERSONAL_NOTIFICATION_FAILED,
+            notification_policy=notification_policy,
+            alert_group=alert_group,
+            reason="Mobile push notification error",
+            notification_step=notification_policy.step,
+            notification_channel=notification_policy.notify_by,
+        )
+
+    device_to_notify = FCMDevice.objects.filter(user=user).first()
+
+    # create an error log in case user has no devices set up
+    if not device_to_notify:
+        _create_error_log_record()
+        logger.error(f"Error while sending a mobile push notification: user {user_pk} has no device set up")
+        return
+
+    message = _get_alert_group_escalation_fcm_message(alert_group, user, device_to_notify, critical)
+    _send_push_notification(device_to_notify, message, _create_error_log_record)
+
+
+def should_we_send_going_oncall_push_notification(
+    now: timezone.datetime, user_settings: "MobileAppUserSettings", schedule_event: ScheduleEvent
+) -> bool:
+    NOTIFICATION_TIMING_BUFFER = 15 * 60  # 15 minutes in seconds
```

Comment:
Since the task is scheduled to run every 10 mins, is it possible we may send 2 notifications for the same user/shift combination?

Reply:
good point 👍 I ended up storing a key in redis for ~80mins to keep track of these notifications ([change](https://github.com/grafana/oncall/pull/1814/commits/c323eac5d1af995f7ac467b1d488bf869d100d06#diff-b967f09465191f5c7bf2d5cca2356b32aeec88ef5d99946128b8edec0a084cb4R377-R429)) and avoid duplicates